includes/validate.php

```diff
@@ -184,65 +184,47 @@ function edac_validate( $post_ID, $post, $action ) {
 function edac_remove_corrected_posts( $post_ID, $type, $pre = 1, $ruleset = 'php' ) {
 	global $wpdb;
 
-	$rules        = edac_register_rules();
-	$js_rule_ids  = array();
-	$php_rule_ids = array();
+	$rules      = edac_register_rules();
+	$rule_slugs = array();
 	foreach ( $rules as $rule ) {
-		if ( array_key_exists( 'ruleset', $rule ) && 'js' === $rule['ruleset'] ) {
-			$js_rule_ids[] = $rule['slug'];
-		} else {
-			$php_rule_ids[] = $rule['slug'];
+		if ( 'js' === $ruleset && isset( $rule['ruleset'] ) && 'js' === $rule['ruleset'] ) {
+			$rule_slugs[] = $rule['slug'];
+		} elseif ( 'js' !== $ruleset ) {
+			$rule_slugs[] = $rule['slug'];
 		}
 	}
-	
-	// Build a sql sanitized list from an array
-	// See: https://stackoverflow.com/questions/10634058/wordpress-prepared-statement-with-in-condition .
-	$js_rule_ids = array_map(
-		function ( $v ) {
-			return "'" . esc_sql( $v ) . "'";
-		},
-		$js_rule_ids
-	);
-	$js_rule_ids = implode( ',', $js_rule_ids );
-
-	// Build a sql sanitized list from an array
-	// See: https://stackoverflow.com/questions/10634058/wordpress-prepared-statement-with-in-condition .
-	$php_rule_ids = array_map(
-		function ( $v ) {
-			return "'" . esc_sql( $v ) . "'";
-		},
-		$php_rule_ids
-	);
-	$php_rule_ids = implode( ',', $php_rule_ids );
-
 
 	if ( 1 === $pre ) {
 
-		// set record flag before validating content.
-		$sql = $wpdb->prepare( 'UPDATE ' . $wpdb->prefix . 'accessibility_checker SET recordcheck = %d WHERE siteid = %d and postid = %d and type = %s', 0, get_current_blog_id(), $post_ID, $type );
-		
-		if ( 'js' === $ruleset ) {
-			$sql = $sql . ' AND rule IN(' . $js_rule_ids . ')';
-		} else {
-			$sql = $sql . ' AND rule IN(' . $php_rule_ids . ')';
-		}
-
-		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching, WordPress.DB.PreparedSQL.NotPrepared -- Prepare used above, Using direct query for adding data to database, caching not required for one time operation.
-		$wpdb->query( $sql );
+		// Set record flag before validating content.
+		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching -- Using direct query for adding data to database, caching not required for one time operation.
+		$wpdb->query(
+			$wpdb->prepare(
+				sprintf(
+					"UPDATE {$wpdb->prefix}accessibility_checker SET recordcheck = %%d WHERE siteid = %%d and postid = %%d and type = %%s AND rule IN (%s)",
+					implode( ',', array_fill( 0, count( $rule_slugs ), '%s' ) )
+				),
+				array_merge(
+					array( 0, get_current_blog_id(), $post_ID, $type ),
+					$rule_slugs
+				)
+			)
+		);
```

Comment:
Breaking down what this does:
* All placeholders are escaped (for example `%%d`), except the last one (`%s`). 
* The `implode( ',', array_fill(...) )` line will take the `$rule_slugs` var, and if it has 3 items it will return `%s,%s,%s`, if it has 4 items it will return `%s,%s,%s,%s` etc.
After `sprintf`, this line:
```php
"UPDATE {$wpdb->prefix}accessibility_checker SET recordcheck = %%d WHERE siteid = %%d and postid = %%d and type = %%s AND rule IN (%s)",
```
gets converted to this (assuming in this example that `$rule_slugs` contains 3 items):
```php
"UPDATE {$wpdb->prefix}accessibility_checker SET recordcheck = %d WHERE siteid = %d and postid = %d and type = %s AND rule IN (%s,%s,%s)",
```
* The `$wpdb->prepare()` method then takes over, and properly quotes `%s` where needed.

The same method is now used in WP-Core to get around issues where it was previously hard to properly prepare queries.